### PR TITLE
`std::context::Context` and usage checks

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

see #558 

### How was it fixed?

Created an `std::context` module and added a `Context` struct. Usage of this struct is guided by the compiler. To see how usage is guided, I recommend looking at the `ctx_` compiler error snapshots.

There are a few things missing in this PR:

- Function access in the Json ABI should take `ctx` into account. This would be a good first issue, so I'll write it up and see if someone picks it up.
- There should be a `std::prelude` module that reexports items like `Context`. I tried to implement this, but ran into some cycle errors and decided to set it aside for now. For now, `Context` must be imported by the user. I'll write an issue for this also.

Note: `ctx` can only be used safely in contract functions. This might be too strict..

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: crates/analyzer/src/db/queries/functions.rs [hunk 1]
```diff
@@ -77,7 +78,28 @@ pub fn function_signature(
                     ))),
                 });
 
-                if let Ok(Some(named_item)) = scope.resolve_name(&name.kind) {
+                if name.kind == "ctx" {
+                    if !function.parent(db).is_contract() {
+                        scope.error(
+                            "ctx can only be used within a contract function",
```
[the remark this answers is not in the document]
Comment (g-r-a-n-t) [reply crates/analyzer/src/db/queries/functions.rs:84]: keep disallowed for now

File: crates/yulgen/src/constructor.rs [hunk 2]
```diff
@@ -47,6 +48,12 @@ pub fn build_with_init(
 
     let deployment = deployment();
 
+    let init_call = if expects_ctx {
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/constructor.rs:51]: comments

File: crates/yulgen/src/db/queries/functions.rs [hunk 3]
```diff
@@ -53,8 +53,14 @@ pub fn function_sig_abi_types(
     let adb = db.upcast();
     let sig = function.signature(adb);
     let return_type = sig.return_type.clone().expect("return type error");
+    let param_types = sig
+        .param_types()
+        .into_iter()
+        .skip(if sig.ctx_decl.is_some() { 1 } else { 0 })
+        .collect::<Vec<_>>();
+    println!("{}: {:?}", function.name(adb), param_types);
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/db/queries/functions.rs:61]: delete

File: crates/yulgen/src/db/queries/functions.rs [hunk 4]
```diff
@@ -53,8 +53,14 @@ pub fn function_sig_abi_types(
     let adb = db.upcast();
     let sig = function.signature(adb);
     let return_type = sig.return_type.clone().expect("return type error");
+    let param_types = sig
+        .param_types()
+        .into_iter()
+        .skip(if sig.ctx_decl.is_some() { 1 } else { 0 })
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/db/queries/functions.rs:59]: `external_param_types`?

File: crates/yulgen/src/mappers/expressions.rs [hunk 5]
```diff
@@ -138,7 +134,13 @@ fn expr_call(context: &mut FnContext, exp: &Node<fe::Expr>) -> yul::Expression {
         CallType::TypeConstructor(Type::Base(Base::Numeric(integer))) => {
             math_operations::adjust_numeric_size(&integer, yul_args[0].clone())
         }
-        CallType::TypeConstructor(_) => yul_args[0].clone(),
+        CallType::TypeConstructor(typ) => {
+            if matches!(typ, Type::Contract(_)) {
+                yul_args[1].clone()
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/mappers/expressions.rs:139]: comment

File: crates/yulgen/src/mappers/functions.rs [hunk 6]
```diff
@@ -145,6 +145,7 @@ fn emit(context: &mut FnContext, stmt: &Node<fe::FuncStmt>) -> yul::Statement {
         let event_values = args
             .kind
             .iter()
+            .skip(1)
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/mappers/functions.rs:148]: comment

File: crates/yulgen/src/runtime/abi_dispatcher.rs [hunk 7]
```diff
@@ -61,7 +70,11 @@ fn dispatch_arm(
     // nothing. Otherwise, we encode the value and return it.
     let call_and_maybe_encode_return = {
         let name = identifier! { (qualified_name) };
-        let call = expression! { [name]([param_exprs...]) };
+        let call = if expects_ctx {
```
Comment (g-r-a-n-t) [inline crates/yulgen/src/runtime/abi_dispatcher.rs:73]: comment

File: crates/test-files/fixtures/features/ctx_param_external_func_call.fe [hunk 8]
```diff
@@ -0,0 +1,59 @@
+use std::context::Context
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/features/ctx_param_external_func_call.fe:1]: delete this and rename the other one

**Review by sbillig — APPROVED** [review]
Looks good!

**Pull request merged**